lib/hanami/action/flash.rb

```diff
@@ -125,94 +133,124 @@ def inspect
       #
       # @since 0.3.0
       # @api private
-      def flash
+      def _data
         @session[SESSION_KEY] || {}
       end
 
-      # The flash registry that holds the data **only for** the current request
+      # Remove the flash entirely from the session if empty.
       #
-      # @return [Hash] the flash for the current request
+      # @return [void]
       #
       # @since 0.3.0
       # @api private
-      def data
-        flash[@request_id] || {}
+      #
+      # @see Hanami::Action::Flash#empty?
+      def remove!
+        if empty?
+          @session.delete(SESSION_KEY)
+          @session.delete(KEPT_KEY)
+        end
       end
 
-      # Expire the stale data from the previous request.
+      # Values from session and kept
       #
-      # @return [void]
+      # @return [Hash]
       #
       # @since 0.3.0
       # @api private
-      def expire_stale!
-        flash.each do |request_id, _|
-          flash.delete(request_id) if delete?(request_id)
-        end
+      def _values
+        _data.merge(kept_data)
       end
 
-      # Remove the flash entirely from the session if empty.
+      # Get the kept request data
       #
-      # @return [void]
+      # @return [Array]
       #
-      # @since 0.3.0
+      # @since x.x.x
       # @api private
-      #
-      # @see Hanami::Action::Flash#empty?
-      def remove!
-        @session.delete(SESSION_KEY) if empty?
+      def kept
+        @session[KEPT_KEY] || []
       end
 
-      # @since 1.0.0
-      # @api private
+      # Merge current data into KEPT_KEY hash
+      #
+      # @return [Hash] the current value of KEPT_KEY
       #
-      # @see Hanami::Action::Flash#[]
-      def merge!
-        @merged = last_request_flash.merge(data)
+      # @since x.x.x
+      # @api private
+      def keep_data
+        new_kept_data = kept << { count: 0, data: _data }.to_json
```

Comment:
For consistency with https://github.com/hanami/controller/pull/267/files#diff-ce13f23d4ed57d06a91f48c61b42c054R209 this should be `JSON.generate` instead of using `to_json`.